GitVersionExe/GitPreparer.cs

```diff
@@ -20,31 +21,74 @@ public bool IsDynamicGitRepository
+            var targetPath = CalculateTemporaryRepositoryPath(arguments.TargetUrl);
+            DynamicGitRepositoryPath = CreateDynamicRepository(targetPath, arguments.Authentication, arguments.TargetUrl, arguments.TargetBranch);
+        }
+
+        string CalculateTemporaryRepositoryPath(string targetUrl)
+        {
+            var userTemp = Path.GetTempPath();
+            var repositoryName = targetUrl.Split('/', '\\').Last().Replace(".git", string.Empty);
+            var possiblePath = Path.Combine(userTemp, repositoryName);
+
+            // Verify that the existing directory is ok for us to use
+            if (Directory.Exists(possiblePath))
             {
-                gitPath = GetGitInfoFromUrl();
+                if (!GitRepoHasMatchingRemote(possiblePath, targetUrl))
+                {
+                    var i = 1;
+                    var originalPath = possiblePath;
+                    bool possiblePathExists;
+                    do
+                    {
+                        possiblePath = string.Concat(originalPath, "_", i++.ToString());
+                        possiblePathExists = Directory.Exists(possiblePath);
+                    } while (possiblePathExists && !GitRepoHasMatchingRemote(possiblePath, targetUrl));
+                }
             }
 
-            return GitDirFinder.TreeWalkForGitDir(gitPath);
+            return possiblePath;
         }
 
-        string GetGitInfoFromUrl()
+        static bool GitRepoHasMatchingRemote(string possiblePath, string targetUrl)
         {
-            var gitRootDirectory = Path.Combine(arguments.TargetPath, "_dynamicrepository");
-            var gitDirectory = Path.Combine(gitRootDirectory, ".git");
-            if (Directory.Exists(gitRootDirectory))
+            try
+            {
+                using (var repository = new Repository(possiblePath))
+                {
+                    return repository.Network.Remotes.Any(r => r.Url == targetUrl);
+                }
+            }
+            catch (Exception)
             {
-                Logger.WriteInfo(string.Format("Deleting existing .git folder from '{0}' to force new checkout from url", gitRootDirectory));
+                return false;
+            }
+            
+        }
+
+        public string GetDotGitDirectory()
+        {
+            if (IsDynamicGitRepository)
+                return DynamicGitRepositoryPath;
 
-                DeleteHelper.DeleteGitRepository(gitRootDirectory);
+            return GitDirFinder.TreeWalkForDotGitDir(arguments.TargetPath);
+        }
+
+        static string CreateDynamicRepository(string targetPath, Authentication authentication, string repositoryUrl, string targetBranch)
+        {
+            var gitDirectory = Path.Combine(targetPath, ".git");
+            if (Directory.Exists(targetPath))
+            {
+                Logger.WriteInfo("Git repository already exists at {0}, skipping clone");
```

Comment:
The normalise git repo step later will fetch/normalise for us? This side of GitVersion I'm not too strong on though.

Sent from my Windows Phone
________________________________
From: Geert van Horrik<mailto:notifications@github.com>
Sent: ý22/ý03/ý2015 13:36
To: ParticularLabs/GitVersion<mailto:GitVersion@noreply.github.com>
Cc: Jake Ginnivan<mailto:jake@ginnivan.net>
Subject: Re: [GitVersion] Dynamic repository location (#399)


In GitVersionExe/GitPreparer.cs<https://github.com/ParticularLabs/GitVersion/pull/399#discussion_r26903087>:

>
> -                DeleteHelper.DeleteGitRepository(gitRootDirectory);
> +            return GitDirFinder.TreeWalkForDotGitDir(arguments.TargetPath);
> +        }
> +
> +        static string CreateDynamicRepository(string targetPath, Authentication authentication, string repositoryUrl, string targetBranch)
> +        {
> +            var gitDirectory = Path.Combine(targetPath, ".git");
> +            if (Directory.Exists(targetPath))
> +            {
> +                Logger.WriteInfo("Git repository already exists at {0}, skipping clone");


Clone should be skipped, but shouldn't we be doing a fetch / pull?

—
Reply to this email directly or view it on GitHub<https://github.com/ParticularLabs/GitVersion/pull/399/files#r26903087>.